src/main/java/bio/terra/service/resourcemanagement/google/GoogleResourceService.java

```diff
@@ -68,44 +77,148 @@ public GoogleResourceService(
+        logger.info("metadataCreationFailed = " + metadataCreationFailed);
+        if (metadataCreationFailed) {
+            // insert failed. lookup the existing bucket_resource row
+            googleBucketResource = resourceDao.getBucket(bucketRequest);
+        }
+
+        // this fault is used by the ResourceLockTest
+        if (configService.testInsertFault(ConfigEnum.BUCKET_LOCK_CONFLICT_STOP_FAULT)) {
+            try {
+                logger.info("BUCKET_LOCK_CONFLICT_STOP_FAULT");
+                while (!configService.testInsertFault(ConfigEnum.BUCKET_LOCK_CONFLICT_CONTINUE_FAULT)) {
+                    logger.info("Sleeping for CONTINUE FAULT");
+                    TimeUnit.SECONDS.sleep(5);
+                }
+                logger.info("BUCKET_LOCK_CONFLICT_CONTINUE_FAULT");
+            } catch (InterruptedException ex) {
+                Thread.currentThread().interrupt();
+                throw new LoadLockFailureException("Unexpected interrupt during bucket lock fault");
             }
-        } catch (GoogleResourceNotFoundException e) {
-            logger.info("no bucket resource metadata found for project: {}", projectResource.getGoogleProjectId());
         }
 
-        // the bucket might already exist even though we don't have metadata for it
-        Bucket bucket = getBucket(bucketRequest.getBucketName()).orElseGet(() -> newBucket(bucketRequest));
+        String bucketName = bucketRequest.getBucketName();
+        String lockingFlightId = googleBucketResource.getFlightId();
+        logger.info("lockingFlightId = " + lockingFlightId);
+        Bucket bucket;
+        if (lockingFlightId == null) {
+            // GET case. the row in the bucket_resource table is unlocked. the bucket has already been created
+
+            // throw an exception if the bucket does not already exist
+            bucket = getBucket(bucketName)
+                .orElseThrow(() -> new GoogleResourceNotFoundException("Bucket not found: " + bucketName));
+        } else if (lockingFlightId.equals(flightId)) {
+            // CREATE case. the row in the bucket_resource table is locked by this flight. we need to create it here
+
+            // check if the bucket already exists
+            Optional<Bucket> existingBucket = getBucket(bucketName);
+            if (!existingBucket.isPresent()) {
+                // bucket DOES NOT EXIST. create it and unlock
+                bucket = newBucket(bucketRequest);
+                resourceDao.unlockBucket(bucketName, flightId);
+            } else {
+                // bucket EXISTS. if this is a recovery flight or testing environment, use it. otherwise throw exception
+                boolean springProfileIsForTesting = Arrays.stream(springEnvironment.getActiveProfiles()).anyMatch(
+                    springProfile -> springProfile.equals("integration") || springProfile.equals("connectedtest"));
+                if (metadataCreationFailed || springProfileIsForTesting) {
```

Comment:
I added this check here to handle the case where the metadata row already exists and is locked by the current flight AND the bucket has already been created. In that case, I don't want to throw an exception regardless of the application property. I was thinking this could happen if a step failed before writing its state, then got picked up and re-run.